src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -0,0 +1,323 @@
+ucc_tl_ucp_allreduce_sliding_window_req_test(ucs_status_ptr_t request,
+                                             ucc_tl_ucp_task_t *task)
+{
+    if (request == NULL) {
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(request)) {
+        tl_error(UCC_TASK_LIB(task), "unable to complete UCX request=%p: %d\n",
+                 request,
+                 UCS_PTR_STATUS(request));
+        return UCS_PTR_STATUS(request);
+    } else {
+        return ucp_request_check_status(request);
+    }
+}
+
+void ucc_tl_ucp_allreduce_sliding_window_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t                  *task           = ucc_derived_of(coll_task,
+                                                                        ucc_tl_ucp_task_t);
+    ucc_rank_t                          size           = (ucc_rank_t)task->subset.map.ep_num;
+    ucc_datatype_t                      dtype          = TASK_ARGS(task).dst.info.datatype;
+    size_t                              dt_size        = ucc_dt_size(dtype);
+    uint32_t                            host_team_size = size;
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe           = task->allreduce_sliding_window.pipe;
+    ucc_tl_ucp_allreduce_sw_buf_t      *accbuf         = &pipe->accbuf;
+    ucp_request_param_t                 req_param      = {0};
+    int                                 i              = 0;
+
+    size_t                         remaining_elems;
+    size_t                         get_idx;
+    size_t                         count;
+    size_t                         get_offset;
+    size_t                         data_size;
+    int                            src_rank;
+    void                          *src_addr;
+    void                          *dst_addr;
+    ucs_status_ptr_t               request;
+    size_t                         red_idx;
+    ucc_tl_ucp_allreduce_sw_buf_t *redbuf;
+    ucc_tl_ucp_allreduce_sw_buf_t *getbuf;
+    size_t                         put_offset;
+    int                            window;
+    int                            put_idx;
+    int                            dst_rank;
+
+    if (pipe->count_serviced < pipe->my_count) {
+        if ((pipe->count_received < pipe->my_count)
+            && (pipe->done_get < host_team_size)
+            && (pipe->avail_buffs > 0)
+            && (accbuf->state != REDUCED && accbuf->state != SENDING))
+        {
+            remaining_elems = pipe->my_count - pipe->count_received;
+            get_idx         = pipe->get_idx % pipe->num_buffers;
+            count           = ucc_min(pipe->buffer_size/dt_size,
+                                      remaining_elems);
+            get_offset      = pipe->count_received * dt_size
+                                + pipe->my_offset;
+            data_size       = count * dt_size;
+            src_rank        = pipe->src_rank;
+            getbuf          = accbuf->state == FREE ? 
+                                accbuf : &pipe->getbuf[get_idx];
```

Comment:
What do you think about using a memory pool (see mpool) to dynamically request buffers from, instead of doing manually like here?

Reply:
Sorry, can you please point me to where `pipe->getbuf` is allocated?

Reply:
The buffers are allocated in the UROM ucc worker here: https://gitlab-master.nvidia.com/hpc_sw_arch/urom/-/merge_requests/148/diffs#c81a79030f411e05cc204356423a4277e4216c04_0_58

Then, the pointers to the buffers are set and passed in via the global work buffer argument on the coll_args.

I am open to the idea of using a memory pool for these buffers. What would be the advantage? Right now, the UROM ucc worker allocates the buffers. It is reconfigurable--with a UROM command it can free and then allocate a different count of buffers with another size.